spec.md

```diff
@@ -1904,9 +1903,11 @@ message NodePublishVolumeRequest {
   // The path to which the volume will be published. It MUST be an
   // absolute path in the root filesystem of the process serving this
   // request. The CO SHALL ensure uniqueness of target_path per volume.
-  // The CO SHALL ensure that the process serving the request has
-  // `read` and `write` permissions to the path, and is able to create
-  // files or directories at the path if it does not exist.
+  // The CO SHALL ensure that the path exists and that the process
+  // serving the request has `read` and `write` permissions to the path.
+  // For volumes with an access type of block, the path MUST be a file.
+  // For volumes with an access type of mount, the path MUST be a
+  // directory. The SP SHALL bind mount the volume to the target path.
```

Comment:
do we really have to say "bind mount" here? Isn't it enough to say "mount"?

Reply:
@jdef The problem I think many people have is that the word "mount" traditionally implies what happens when you add a new filesystem into the VFS at some directory. For block volumes, we've opted to actually bind mount a single device file. There are lots of alternatives we could have chosen, including symlinks or returning a string. Presumably there is some advantage to the bind mount approach (I wasn't around when this decision was made) or perhaps the spec was left so vague that Kubernetes just chose to use bind mounts rather than the other alternatives and that became the de-facto way to do it. This kind of open-endedness is poison for specifications because it makes it very easy for 2 different implementation that both conform to the spec to not interoperate with eachother.

I could change the wording to make it clear that the bind mount is only a requirement for block access types, and that an ordinary mount is find for mount (filesystem) access types.

Reply:
OK, I understand where you're coming from. Again, this is where the proposal gets a bit too specific w/ it's *requirements* of a plugin. I'd like to see just a tiny bit of flexibility here.

Can you replace "SHALL bind mount" with "SHOULD bind mount"?

Reply:
FWIW there's multiple ways to interpret "bind mount" so I'm not sure that there's too much actually gained by keeping the word "bind".

Fun read: https://unix.stackexchange.com/questions/198590/what-is-a-bind-mount

Reply:
I'm still having a hard time getting behind this change request for `NodePublishVolume`. If a plugin doesn't support node-stage and simply wants to `mknod` a block device at `target_path` then this proposal blocks that behavior. Forcing a plugin to `mknod` elsewhere and then to "bind mount" it to a file that the CO is creating here seems like a mistake.

I think we should keep more neutral wording here. The CO SHALL NOT create a file system object at `target_path`. Plugin implementation recommendations should be made elsewhere in the spec and *should* avoid *requirement* wording like SHALL, and MUST.

Reply:
The word SHOULD implies to me that you can not do something and then there are still some scenarios where that can lead to a good result. SPs must do something at the target path to ensure a good result. We owe it to SP implementors to either enumerate all the valid things an SP can do or describe the end result in such a way that at least a few of the valid approaches are obvious.

Reply:
I'm with you on the fact that there are valid alternatives that can work. Doing a mknod at the target path makes perfect sense to me, but then we have to change the wording to say the CO MUST NOT create anything there, which is is the opposite of what Kubernetes has done and will require changing Kubernetes again (not impossible, just annoying).

Reply:
I could go either way on this, @saad-ali do you feel strongly? The one outcome I don't want is to leave it unspecified which side should create the file/device because then a de-facto standard of "whatever works with kubernetes" will emerge.

Reply:
I have a slight preference to leave the spec lose enough for alternate implementations.

> We owe it to SP implementors to either enumerate all the valid things an SP can do or describe the end result in such a way that at least a few of the valid approaches are obvious.

I'd prefer the later.